extractors/reddit/reddit.go

```diff
@@ -0,0 +1,190 @@
+	html, err := request.Get(url, referer, nil)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	// set thread number to 1 manually to avoid http 412 error
+	option.ThreadNumber = 1
+
+	var fileType = ""
+	contentName := utils.MatchOneOf(html, `<title>(.+?)<\/title>`)[1]
+	if utils.MatchOneOf(html, `meta property="og:video" content=.*HLSPlaylist`) != nil {
+		fileType = "mp4"
+	} else if utils.MatchOneOf(html, `<meta property="og:type" content="image"/>`) != nil {
+		fileType = "img"
+	} else if utils.MatchOneOf(html, `https:\/\/preview\.redd\.it\/.*gif`) != nil {
+		fileType = "gif"
+	}
+
+	if fileType == "mp4" {
+		mp4URL := utils.MatchOneOf(html, `https://v.redd.it/(.+?)/HLSPlaylist`)[1]
+		if mp4URL == "" {
+			return nil, errors.New("can't match mp4 content downloadable url")
+		}
+
+		audioURL := fmt.Sprintf("%s%s%s", redditMP4API, mp4URL, audioURLPart)
+		size, err := request.Size(audioURL, referer)
+		audioPart := &extractors.Part{
+			URL:  audioURL,
+			Size: size,
+			Ext:  "mp3",
+		}
+
+		resMap := map[string]string{
+			"720p": "/DASH_720.mp4",
+			"480p": "/DASH_480.mp4",
+			"360p": "/DASH_360.mp4",
+			"240p": "/DASH_240.mp4",
+			"220p": "/DASH_220.mp4",
+		}
+		streams := make(map[string]*extractors.Stream, len(resMap))
+		for res, urlparts := range resMap {
+			if err != nil {
+				return nil, err
+			}
+			resURL := fmt.Sprintf("%s%s%s", redditMP4API, mp4URL, urlparts)
+			size, err := request.Size(resURL, referer)
+			if err != nil {
+				return nil, err
+			}
+			if err != nil {
+				return nil, err
+			}
+			parts := make([]*extractors.Part, 0, 2)
+			parts = append(parts, &extractors.Part{
+				URL:  resURL,
+				Size: size,
+				Ext:  "mp4",
+			})
+			parts = append(parts, audioPart)
+			var totalSize int64
+			for _, part := range parts {
+				totalSize += part.Size
+			}
+			streams[res] = &extractors.Stream{
+				Parts:   parts,
+				Size:    totalSize,
+				Quality: res,
+				NeedMux: true,
+			}
+		}
+
+		return []*extractors.Data{
+			{
+				Site:    "Reddit reddit.com",
+				Title:   contentName,
+				Type:    extractors.DataTypeVideo,
+				Streams: streams,
+				URL:     url,
+			},
+		}, nil
+	} else if fileType == "gif" {
+		gifURL := utils.MatchOneOf(html, `https:\/\/preview\.redd\.it\/.*?\.gif\?format=mp4.*?"`)[0]
+		if gifURL == "" {
+			return nil, errors.New("can't match gif content downloadable url")
+		}
+
+		gifURL = strings.ReplaceAll(gifURL, "&amp;", "&")
+		gifURL = strings.ReplaceAll(gifURL, "\"", "")
+
+		videoSize, err := request.Size(gifURL, "reddit.com")
+		if err != nil {
+			return nil, errors.New("can't get video size")
+		}
+		contentData := make([]*extractors.Part, 0)
+		contentData = append(contentData, &extractors.Part{
+			URL:  gifURL,
+			Size: videoSize,
+			Ext:  "gif",
+		})
+
+		streams := map[string]*extractors.Stream{
+			"default": {
+				Parts: contentData,
+				Size:  videoSize,
+			},
+		}
```

Comment:
You don't have to define a `contentData` variable

```suggestion
		streams := map[string]*extractors.Stream{
			"default": {
				Parts: []*extractors.Part{{...}},
				Size:  videoSize,
			},
		}
```